### PR TITLE
Implement authorization by publishing key id.

### PR DESCRIPTION
**Pull request by peterhuene**
This PR implements authorization of publishing operations by the key id used to sign the record being published.

By default, the server `Config` denies all keys, and keys can be specified per-namespace with the `with_authorized_key` method.

Authorization can also be disabled (allow any key id to claim a package name) with the `with_no_authorization` method.

A consequence of this change is that package names are now restricted to valid package names in a component model ID (e.g. `wasi:http`).

The tests have also been refactored to better share the test implementations between the in-memory and postgres data stores.

In the future, the authorization scheme could be further refined to specific package names rather than entire namespaces.

### CODE REVIEW

File: crates/server/src/api/mod.rs
Comment (Kylebrown9) [inline crates/server/src/api/mod.rs:27]: At some point I would like policies to be a `Vec<dyn Policy>`-style construction where this namespace control policy isn't special.

In this flow, the `Policy` trait would either
* have methods for each of the possible things it could have an opinion on,
* would have getters/factories that optionally return sub-policies for specific areas,
* or would actually be an `enum` such that each policy refers to a single area and you just add multiple policies to the list if you need to refer to multiple areas.

File: crates/server/src/api/v1/package.rs
Comment (Kylebrown9) [inline crates/server/src/api/v1/package.rs:109]: I think this belongs in the protocol crate next to `LogId` and we'll want to make `LogId` derivable from this.

File: crates/server/src/datastore/memory.rs
Comment (Kylebrown9) [inline crates/server/src/datastore/memory.rs:582]: What is this function doing?

Comment (peterhuene) [reply crates/server/src/datastore/memory.rs:582]: We need to verify the signature of the record being published prior to accepting the record for processing; to do that, the data store needs to fetch the package's validation state to determine the public key to verify with.

This prevents people from sending us requests where the `keyId` property is explicitly set to an acceptable (known) key id, as these are not private, and then giving us a record signed with a mismatched private key.

By doing so, it preemptively rejects the record before it ever makes it into the database for later validation; validation would have rejected the record anyway due to the invalid signature.

As authorization is based on these public key ids, we shouldn't allow unauthorized keys to write to the database in any form.

File: crates/server/src/lib.rs
Comment (peterhuene) [reply crates/server/src/lib.rs:143]: I'll take a stab at moving this to a proper "record policy" trait with a particular implementation for authorizing.

File: crates/api/src/v1/package.rs
Comment (peterhuene) [reply crates/api/src/v1/package.rs:118]: I can standardize the error types in the other direction instead (there were other tuple variants and this one was the outlier) and make them all struct variants.

Comment (Kylebrown9) [reply crates/api/src/v1/package.rs:118]: I guess if it's specifically a convention that "error enums with a single message field are mono-tuples" that makes sense to me. I just don't want to lose names when they can't be inferred so easily.

Speaking of which, it might make sense to document some conventions at some point.

File: crates/server/src/api/mod.rs
Comment (peterhuene) [reply crates/server/src/api/mod.rs]: I moved this out of `Config` and into an `AuthorizedKeyPolicy` implementing the logic.

The mechanism is minimal right now, but we can expand on the design in future work.

File: crates/server/src/lib.rs
Comment (peterhuene) [reply crates/server/src/lib.rs]: I moved this to the specific policy and `Config` now only has a method to set the record policy to use.

File: crates/server/src/lib.rs
Comment (peterhuene) [reply crates/server/src/lib.rs:26]: Agreed, `wit_parser` should probably expose this; right now this (and the "interface id" parsing) is crate internal.

Comment (peterhuene) [reply crates/server/src/lib.rs:26]: Actually, I think we can use `KebabStr` from `wasmparser` for this validation. I'll fix.

File: crates/client/src/lib.rs
Comment (Kylebrown9) [inline crates/client/src/lib.rs:240]: Do we use this in non-slice contexts now?

File: crates/server/src/policy/record/authorization.rs
Comment (Kylebrown9) [inline crates/server/src/policy/record/authorization.rs:11]: Thoughts on making this `Serialize + Deserialize` so we can pull it from a config file?

File: crates/protocol/src/registry.rs
Comment (Kylebrown9) [inline crates/protocol/src/registry.rs:97]: Should this be called `try_new` and/or does it even need to exist outside the parse & serde impls?

File: crates/server/src/datastore/memory.rs
Comment (Kylebrown9) [inline crates/server/src/datastore/memory.rs:214]: why/how is this unused?

**Review by Kylebrown9 — APPROVED** [review]
LGTM

I have a few questions/suggestions, but I'll pre-approve so once addressed you can merge without waiting for me again.

File: crates/server/src/datastore/memory.rs
Comment (peterhuene) [reply crates/server/src/datastore/memory.rs:214]: The in-memory data store doesn't keep the names around, but the postgres one does (really only so one can look at the actual SQL table and see the package ids are; they aren't read out by the server).

Until we have a query API, it really helps the operator to figure out what logs are present.

File: crates/server/src/policy/record/authorization.rs
Comment (peterhuene) [reply crates/server/src/policy/record/authorization.rs:11]: I think it'd be best addressed in a follow-up PR to properly represent a server configuration as a single file that can be deserialized to enable the policies (with their specific configurations).

I purposely punted on that (i.e. there's no way to enable policies in the server CLI) as I would want to have a proper discussion around configuration file format.

File: crates/client/src/lib.rs
Comment (peterhuene) [reply crates/client/src/lib.rs:240]: We do in the tests currently. As this method is only iterating on the package ids with no requirement it needs to be a slice, it actually helps consumers like `cargo-component` that are iterating on some other data structure and projecting into it to get the warg package identifiers; this change removes the need for callers to collect into a contiguous collection.

File: crates/protocol/src/registry.rs
Comment (peterhuene) [reply crates/protocol/src/registry.rs:97]: This exists for anyone that wants to move the string into the `PackageId`; in low-level libraries like this, I generally like to give people the option of not cloning if possible.

Happy to call it `try_new` (I feel like the naming around fallible a `new` isn't consistent in a lot of codebases, though; `wasmtime` has a bunch of `new` that return `T` and also `Result<T>`).

Comment (lann) [reply crates/protocol/src/registry.rs:97]: Yeah fallible `new` is quite common.

**Pull request merged**
